File: backend/app/utils/encryption.py
"""Utilidades de cifrado híbrido para peticiones API.

El frontend cifra el payload con AES-256-GCM usando una clave efímera por
petición. Esa clave se protege con la clave pública del backend usando RSA-OAEP.
De esta forma no existe un secreto reutilizable dentro del bundle del navegador.
"""

import json
import os
from base64 import b64decode, b64encode

from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import padding, rsa
from cryptography.hazmat.primitives.ciphers.aead import AESGCM


_SERVER_PRIVATE_KEY = rsa.generate_private_key(public_exponent=65537, key_size=2048)
_SERVER_PUBLIC_KEY_PEM = (
    _SERVER_PRIVATE_KEY.public_key()
    .public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    .decode("utf-8")
)


def get_server_public_key_pem() -> str:
    """Devuelve la clave pública del backend en formato PEM."""

    return _SERVER_PUBLIC_KEY_PEM


def get_server_public_key_meta() -> dict:
    """Devuelve metadatos de la clave pública: PEM y un `kid` (SHA256 hex).

    Retorna un dict con `pem` y `kid`. Esto permite a los clientes verificar
    versiones de clave sin romper el endpoint que ya devuelve solo el PEM.
    """
    # calcular un kid sencillo como SHA-256 sobre el PEM (hex)
    digest = hashes.Hash(hashes.SHA256())
    digest.update(_SERVER_PUBLIC_KEY_PEM.encode("utf-8"))
    kid = digest.finalize().hex()
    return {"pem": _SERVER_PUBLIC_KEY_PEM, "kid": kid}


def decrypt_session_key(encrypted_session_key: str) -> bytes:
    """Desencripta la clave AES efímera usando la clave privada del backend."""

    try:
        encrypted_key_bytes = b64decode(encrypted_session_key)
        return _SERVER_PRIVATE_KEY.decrypt(
            encrypted_key_bytes,
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None,
            ),
        )
    except Exception as e:
        raise ValueError(f"Error desencriptando clave de sesión: {str(e)}")


def encrypt_payload(payload: dict, session_key: bytes) -> dict:
    """Encripta un payload JSON usando AES-256-GCM con una clave efímera."""

    try:
        plaintext = json.dumps(payload).encode("utf-8")
        iv = os.urandom(12)
        cipher = AESGCM(session_key)
        ciphertext = cipher.encrypt(iv, plaintext, None)

        return {
            "iv": b64encode(iv).decode("utf-8"),
            "ciphertext": b64encode(ciphertext).decode("utf-8"),
        }
    except Exception as e:
        raise ValueError(f"Error encriptando payload: {str(e)}")


def decrypt_payload(encrypted_data: dict, session_key: bytes) -> dict:
    """Desencripta un payload AES-256-GCM usando la clave efímera."""

    try:
        iv = b64decode(encrypted_data["iv"])
        ciphertext = b64decode(encrypted_data["ciphertext"])
        cipher = AESGCM(session_key)
        plaintext = cipher.decrypt(iv, ciphertext, None)
        return json.loads(plaintext.decode("utf-8"))
    except Exception as e:
        raise ValueError(f"Error desencriptando payload: {str(e)}")

File: widget/src/utils/encryption.js
/**
 * Convierte un ArrayBuffer a base64 string
 */
function arrayBufferToBase64(buffer) {
  const bytes = new Uint8Array(buffer);
  let binary = '';
  for (let i = 0; i < bytes.byteLength; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

/**
 * Convierte un base64 string a ArrayBuffer
 */
function base64ToArrayBuffer(base64) {
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes.buffer;
}

/**
 * Convierte texto PEM a ArrayBuffer.
 */
function pemToArrayBuffer(pem) {
  const base64 = pem
    .replace(/-----BEGIN PUBLIC KEY-----/g, '')
    .replace(/-----END PUBLIC KEY-----/g, '')
    .replace(/\s+/g, '');

  return base64ToArrayBuffer(base64);
}

/**
 * Genera una clave AES efímera de 32 bytes.
 */
export function generateSessionKey() {
  return crypto.getRandomValues(new Uint8Array(32));
}

/**
 * Importa una clave pública RSA-OAEP en formato PEM.
 */
export async function importPublicKey(publicKeyPem) {
  const keyBuffer = pemToArrayBuffer(publicKeyPem);

  return crypto.subtle.importKey(
    'spki',
    keyBuffer,
    {
      name: 'RSA-OAEP',
      hash: 'SHA-256',
    },
    false,
    ['encrypt']
  );
}

/**
 * Encripta una clave AES con la clave pública del backend.
 */
export async function encryptSessionKey(sessionKey, publicKeyPem) {
  const publicKey = await importPublicKey(publicKeyPem);
  const encryptedKey = await crypto.subtle.encrypt(
    { name: 'RSA-OAEP' },
    publicKey,
    sessionKey
  );

  return arrayBufferToBase64(encryptedKey);
}

/**
 * Encripta un payload JSON usando AES-256-GCM
 * @param {Object} payload - Objeto a encriptar
 * @param {ArrayBufferView|ArrayBuffer} sessionKey - Clave AES efímera (32 bytes)
 * @returns {Promise<Object>} { iv, ciphertext } en formato base64
 */
export async function encryptPayload(payload, sessionKey) {
  try {
    const plaintext = JSON.stringify(payload);
    const plaintextBuffer = new TextEncoder().encode(plaintext);

    const keyMaterial = sessionKey instanceof ArrayBuffer ? sessionKey : sessionKey.buffer;

    const key = await crypto.subtle.importKey(
      'raw',
      keyMaterial,
      { name: 'AES-GCM' },
      false,
      ['encrypt']
    );
    const iv = crypto.getRandomValues(new Uint8Array(12));
    const ciphertext = await crypto.subtle.encrypt(
      {
        name: 'AES-GCM',
        iv: iv
      },
      key,
      plaintextBuffer
    );

    return {
      iv: arrayBufferToBase64(iv),
      ciphertext: arrayBufferToBase64(ciphertext)
    };
  } catch (error) {
    console.error('Error encriptando payload:', error);
    throw new Error('Fallo en encriptación: ' + error.message);
  }
}

/**
 * Desencripta un payload encriptado con AES-256-GCM
 * @param {Object} encryptedData - { iv, ciphertext }
 * @param {ArrayBufferView|ArrayBuffer} sessionKey - Clave AES efímera (32 bytes)
 * @returns {Promise<Object>} Payload desencriptado
 */
export async function decryptPayload(encryptedData, sessionKey) {
  try {
    const { iv, ciphertext } = encryptedData;
    const ivBuffer = base64ToArrayBuffer(iv);
    const ciphertextBuffer = base64ToArrayBuffer(ciphertext);

    const keyMaterial = sessionKey instanceof ArrayBuffer ? sessionKey : sessionKey.buffer;

    const key = await crypto.subtle.importKey(
      'raw',
      keyMaterial,
      { name: 'AES-GCM' },
      false,
      ['decrypt']
    );
    const plaintextBuffer = await crypto.subtle.decrypt(
      {
        name: 'AES-GCM',
        iv: ivBuffer
      },
      key,
      ciphertextBuffer
    );
    const plaintext = new TextDecoder().decode(plaintextBuffer);
    return JSON.parse(plaintext);
  } catch (error) {
    console.error('Error desencriptando payload:', error);
    throw new Error('Fallo en desencriptación: ' + error.message);
  }
}
